test/Microsoft.AspNetCore.Server.KestrelTests/ChunkedRequestTests.cs

```diff
@@ -0,0 +1,294 @@
+        [MemberData(nameof(ConnectionFilterData))]
+        [FrameworkSkipCondition(RuntimeFrameworks.Mono, SkipReason = "Test hangs after execution on Mono.")]
+        public async Task RequestBodyIsConsumedAutomaticallyIfAppDoesntConsumeItFully(ServiceContext testContext)
+        {
+            using (var server = new TestServer(async httpContext =>
+            {
+                var response = httpContext.Response;
+                var request = httpContext.Request;
+
+                Assert.Equal("POST", request.Method);
+
+                response.Headers.Clear();
+                response.Headers["Content-Length"] = new[] { "11" };
+
+                await response.Body.WriteAsync(Encoding.ASCII.GetBytes("Hello World"), 0, 11);
+            }, testContext))
+            {
+                using (var connection = new TestConnection(server.Port))
+                {
+                    await connection.SendEnd(
+                        "POST / HTTP/1.1",
+                        "Content-Length: 5",
+                        "",
+                        "HelloPOST / HTTP/1.1",
+                        "Transfer-Encoding: chunked",
+                        "",
+                        "C", "HelloChunked",
+                        "0",
+                        "",
+                        "POST / HTTP/1.1",
+                        "Content-Length: 7",
+                        "",
+                        "Goodbye");
+                    await connection.ReceiveEnd(
+                        "HTTP/1.1 200 OK",
+                        "Content-Length: 11",
+                        "",
+                        "Hello WorldHTTP/1.1 200 OK",
+                        "Content-Length: 11",
+                        "",
+                        "Hello WorldHTTP/1.1 200 OK",
+                        "Content-Length: 11",
+                        "",
+                        "Hello World");
+                }
+            }
+        }
+
+        [ConditionalTheory]
+        [MemberData(nameof(ConnectionFilterData))]
+        [FrameworkSkipCondition(RuntimeFrameworks.Mono, SkipReason = "Test hangs after execution on Mono.")]
+        public async Task TrailingHeadersAreParsed(ServiceContext testContext)
+        {
+            var requestCount = 10;
+            var requestsReceived = 0;
+
+            using (var server = new TestServer(async httpContext =>
+            {
+                var response = httpContext.Response;
+                var request = httpContext.Request;
+
+                var buffer = new byte[200];
+
+                Assert.Equal(string.Empty, request.Headers["X-Trailer-Header"]);
+
+                while(await request.Body.ReadAsync(buffer, 0, buffer.Length) != 0)
+                {
+                    // read to end
+                }
+
+                if (requestsReceived < requestCount)
+                {
+                    Assert.Equal(new string('a', requestsReceived), request.Headers["X-Trailer-Header"]);
```

Comment:
Might be a stretch even for me :wink: 

Linked to discussion from "Web Hypertext Application Technology Working Group" (whatwg) in https://github.com/aspnet/KestrelHttpServer/issues/622 - think it needs a custom client rather than browser.